Add offer + request options to invites

dborkan opened the pull request:
Add offer + request options to invites.

When generating the invite there are 2 cases:
* For GitHub, we know the userId we are inviting, so now if offer or request is checked, we create a RemoteUser object for that userId, and init the correct permissions.
* For other networks, we don't know the userId(s) who will be receiving the invite, so we add permission token data to the invite URL.

When receiving an invite URL with permission data, the receiver:
* Creates a new RemoteUser object for the user who generated the invite
* Creates a new RemoteInstance object for the instance who generated the invite, and simulates a InstanceHandshake message from that instance with the appropriate consent bits set (so they appear in the buddylist as having offered or requested access).
* If the instance is currently online, it sends the permission code back to the inviter to validate, in a PERMISSION_TOKEN message.
* If the instance is currently offline, it sets ```unusedPermissionToken``` so that it will send the permission code back later when that instance appears online

When the inviter receives a PERMISSION_TOKEN message it:
* Checks if that permission token is still valid for the receiving user
* Sets appropriate consent fields
* Sends a new InstanceHandshake back to the invitee.

If someone tries to spoof an invite (using an invalid permission token or lying in the isOffering field), they will trick the UI of their own uProxy to think that they have permission, but because the other side will not validate their permission token, they will not be able to get access.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2358)
<!-- Reviewable:end -->

lucyhe commented:
Small nits mostly -- looks great + sorry for the delay!

---

Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/generic_core/local-instance.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF5OyWM1xa8s2jm1E55:-KF5OyWM1xa8s2jm1E56:1544156396) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/local-instance.ts#L91)):*
Do you mean that LocalInstanceState doesn't include invitePermissionTokens because we don't want to save that to storage? Maybe the comment can be "returns all fields of LocalInstanceState plus invitePermissionTokens"

---

*[src/generic_core/local-instance.ts, line 108 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF5OrZccT_fwQyzEI7m:-KF5OrZccT_fwQyzEI7n:2094862536) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/local-instance.ts#L108)):*
Do we need to set it to [] if it's undefined?

---

*[src/generic_core/remote-user.ts, line 556 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF6_a8WifmLZyruimkm:-KF6_a8WifmLZyruimkn:-851268660) ([raw file](https://github.com/uproxy/uproxy/blob/94d6fa8e25d57d96fb05237dee635df9f8163c33/src/generic_core/remote-user.ts#L556)):*
Just to check: are we taking these out because https://github.com/uProxy/uproxy/issues/1510 is resolved?

---

*[src/generic_core/social.ts, line 690 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KFBbG9dhtr8Pq9ceBCu:-KFBbG9eogfBmX5gZoge:1132479057) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/social.ts#L690)):*
Maybe add comment: "Quiver, FB, Gmail and Cloud invites do not specify a recipient, so userId will not be defined."

---

*[src/generic_ui/polymer/network-invite-user.ts, line 14 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KFCKK_3Duhmn5ccXkVi:-KFCKK_4gTIgtGiXWn4e:-159923013) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_ui/polymer/network-invite-user.ts#L14)):*
nit: use camelcase: createInviteArgs

---

*[src/generic_ui/polymer/network-invite-user.ts, line 143 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KFCPARy0bpkk21xtl6c:-KFCPARy0bpkk21xtl6d:2066654010) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_ui/polymer/network-invite-user.ts#L143)):*
nit: newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-:-KFCSp4FZNsqA7Mb6T4b:-1241943533)*
<!-- Sent from Reviewable.io -->

dborkan commented:
Review status: 15 of 17 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/generic_core/local-instance.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF5OyWM1xa8s2jm1E55:-KFF6ho59jR_Fm3N5SkI:-797122543) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/local-instance.ts#L91)):*
No, I mean that LocalInstanceState contains 6 fields:
```
  instanceId       :string;
  userId           :string;
  userName         :string;
  imageData        :string;
  invitePermissionTokens :{ [token :string] :PermissionTokenInfo };
  exchangeInviteToken : (token :string, userId :string) => PermissionTokenInfo;
```
However this method only returns 5 fields (it doesn't return exchangeInviteToken since that is a method, not data, and methods shouldn't be written to storage:
```
  instanceId       :string;
  userId           :string;
  userName         :string;
  imageData        :string;
  invitePermissionTokens :{ [token :string] :PermissionTokenInfo };
```

Although now that I look at it again maybe there is some way I can refactor this so ```exchangeInviteToken``` is not part of LocalInstanceState anymore but is some global standalone method...

---

*[src/generic_core/local-instance.ts, line 108 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF5OrZccT_fwQyzEI7m:-KFF71UFPxIBnwerg-7-:739512828) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/local-instance.ts#L108)):*
That initialization (to ```{}```) is taken care of on line 28

---

*[src/generic_core/remote-user.ts, line 556 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF6_a8WifmLZyruimkm:-KFF7SQHjlsaBB-1ulyk:-1942256464) ([raw file](https://github.com/uproxy/uproxy/blob/94d6fa8e25d57d96fb05237dee635df9f8163c33/src/generic_core/remote-user.ts#L556)):*
Yes - because we've removed GTalk

---

*[src/generic_core/social.ts, line 690 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KFBbG9dhtr8Pq9ceBCu:-KFF7ozMEWcQ4cLvVtG4:-1827922361) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/social.ts#L690)):*
Just updated the comment, I just realized data.userId will be set for cloud (it is actually the "userId" of the cloud server we wish to share, as a user in theory could have multiple servers they could potentially share)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2358)*
<!-- Sent from Reviewable.io -->

lucyhe commented:
Small comment changes then :+1:

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/generic_core/local-instance.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KF5OyWM1xa8s2jm1E55:-KFG8HKwA0m2dW2jhZSH:-1997268908) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_core/local-instance.ts#L91)):*
Oh got it! Maybe edit the comment to say "This does not ... because we don't want" instead of "This does not ... because it does we don't want"
Also maybe add a TODO for moving exchangeInviteToken outside of LocalInstanceState? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-:-KFG8dFFFfKh_PO_j5NY:901499137)*
<!-- Sent from Reviewable.io -->

dborkan commented:
Review status: 16 of 17 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/generic_ui/polymer/network-invite-user.ts, line 14 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2358#-KFCKK_3Duhmn5ccXkVi:-KFG7KcedN_BQUN1ydKW:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/623a6bfeeae28467475175ad23e8abf34fe42ec9/src/generic_ui/polymer/network-invite-user.ts#L14)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2358)*
<!-- Sent from Reviewable.io -->
